### PR TITLE
Filter Predicate

### PR DESCRIPTION
**Pull request by Bogi1001**

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/FilterByDatePredicateParser.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	private final CommandLineParser cmdParser = new CommandLineParser().addArgument(dateTimeArg);
+	@Override
+	public ListCommand parse(String str) throws ParseException{
+		cmdParser.parse(str);
+        final LocalDateTime now = referenceDateTime.orElse(LocalDateTime.now());
+
+		 final LocalDate startDate = dateTimeArg.getDate().orElse(now.toLocalDate());
+	     final LocalTime startTime = dateTimeArg.getTime().orElse(LocalTime.of(23, 59));
+	     final LocalDate endDate = dateTimeArg.getDate().orElse(now.toLocalDate());
+	     final LocalTime endTime = dateTimeArg.getTime().orElse(LocalTime.of(23, 59));
```
Comment (pyokagan) [inline src/main/java/seedu/address/logic/parser/FilterByDatePredicateParser.java:38]: This will make `startDate == endDate` and `startTime == endTime`. From reading your other code, I don't think this will lead to a useful behaviour.

File: src/main/java/seedu/address/logic/parser/Parser.java [hunk 2]
```diff
@@ -12,6 +12,8 @@
      * @throws ParseException if the input string could not be parsed. The ranges of the ParseException will
      * contain the substring ranges of the input string which caused the parsing to fail.
      */
-    T parse(String str) throws ParseException;
+
+T parse(String str) throws ParseException;
+
```
Comment (pyokagan) [inline src/main/java/seedu/address/logic/parser/Parser.java:17]: Not sure what this change is for.

File: src/main/java/seedu/address/model/Model.java [hunk 3]
```diff
@@ -115,4 +115,4 @@
         String getName();
     }
 
-}
+}
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/Model.java:118]: @nightism already mentioned this in your previous PR -- there needs to be a newline here.

File: src/main/java/seedu/address/model/filter/FilterByDatePredicate.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	private boolean equalToMin (EventTask event){
+		LocalDateTime start = event.getStart();
+		return start.isEqual(startTime);
+	}
+
+	private boolean equalToMax (EventTask event){
+		LocalDateTime end = event.getEnd();
+		return end.isEqual(endTime);
+	}
+
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/filter/FilterByDatePredicate.java:97]: All of these `greaterThan`, `lesserThan`, `equalToMin` make the code extremely hard to read and makes it hard for me to verify its correctness.

File: src/main/java/seedu/address/storage/config/JsonConfigStorage.java [hunk 5]
```diff
@@ -82,4 +82,4 @@ public void saveConfig(ReadOnlyConfig config, String filePath) throws IOExceptio
         objectMapper.writeValue(file, config);
     }
 
-}
+}
```
Comment (pyokagan) [inline src/main/java/seedu/address/storage/config/JsonConfigStorage.java:85]: Again, not sure what this change is for.

File: src/test/java/seedu/address/logic/commands/AddDeadlineCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	private DeadlineTask task2;
+	private DeadlineTask task3;
+
+	private final Model model = new ModelManager();
+
+	public void DeadlineTasks(DeadlineTask task1, DeadlineTask task2, DeadlineTask task3){
+		try{
+			task1 = new DeadlineTask("Finish project", UNIX_EPOCH.plusHours(1));
+			task2 = new DeadlineTask("Pay rent", UNIX_EPOCH.plusDays(25));
+			task3 = new DeadlineTask("Book tickets for movie", UNIX_EPOCH.plusYears(10));
```
Comment (pyokagan) [inline src/test/java/seedu/address/logic/commands/AddDeadlineCommandTest.java:23]: Won't this be assigning `task1`, `task2`, `task3` to the _arguments_???

File: src/test/java/seedu/address/logic/commands/AddDeadlineCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+		}
+		catch (IllegalValueException e){
+			throw new AssertionError("this should not happen", e);
+		}
+	}
+
+	AddDeadlineCommand command1 = new AddDeadlineCommand(task1);
+	AddDeadlineCommand command2 = new AddDeadlineCommand(task2);
+	AddDeadlineCommand command3 = new AddDeadlineCommand(task3);
```
Comment (pyokagan) [inline src/test/java/seedu/address/logic/commands/AddDeadlineCommandTest.java:32]: No idea what these are for -- they are not even used.

File: src/test/java/seedu/address/logic/commands/AddDeadlineCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+		assertEquals(1, model.addDeadlineTask(task1));
+		assertEquals(2, model.addDeadlineTask(task2));
+		assertEquals(3, model.addDeadlineTask(task3));
+		final CommandResult result = new AddDeadlineCommand(task1).execute(model);
+		final CommandResult result2 = new AddDeadlineCommand(task2).execute(model);
+		final CommandResult result3 = new AddDeadlineCommand(task3).execute(model);
+		assertEquals(task1, model.getDeadlineTask(1));
+		assertEquals(task2, model.getDeadlineTask(2));
+		assertEquals(task3, model.getDeadlineTask(3));
+	}
```
Comment (pyokagan) [inline src/test/java/seedu/address/logic/commands/AddDeadlineCommandTest.java:45]: This test does not test the functionality correctly -- you are adding 3 tasks, then executing `AddDeadlineCommand` which is supposed to add another 3 tasks, and then you compare with the first 3 tasks again, effectively not even testing `AddDeadlineCommand` at all.

File: src/test/java/seedu/address/ui/CommandBox.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    private void setStyleToIndicateIncorrectCommand() {
+        commandTextField.getStyleClass().add("error");
+    }
+
+    public interface OnCommandResultCallback {
+        void call(CommandResult result);
+    }
+
+}
```
Comment (pyokagan) [inline src/test/java/seedu/address/ui/CommandBox.java:85]: Not sure what are all these changes for...

File: src/test/java/seedu/address/ui/CommandBoxTest.java [hunk 10]
```diff
@@ -1,76 +0,0 @@
...
-    public void incorrectCommand_restoresCommandText() throws ParseException, CommandException, IOException {
-        final String inputCommand = "some command";
-        Mockito.when(logic.execute(inputCommand)).thenThrow(new CommandException("some exception"));
-
-        textField.setText(inputCommand);
-        clickOn(textField).push(KeyCode.ENTER);
-        assertEquals(inputCommand, textField.getText());
-    }
-
-}
```
Comment (pyokagan) [inline src/test/java/seedu/address/ui/CommandBoxTest.java:76]: Why did you delete very useful tests?

File: src/test/java/seedu/address/ui/DeadlineTaskListCard.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+                getRoot().getStyleClass().add("finished");
+            } else if (deadlineTask.getDue().isBefore(LocalDateTime.now())) {
+                getRoot().getStyleClass().add("overdue");
+            }
+        } else {
+            getRoot().setVisible(false);
+        }
+    }
+
+}
```
Comment (pyokagan) [inline src/test/java/seedu/address/ui/DeadlineTaskListCard.java:44]: Why is this moved into the test directory?

File: src/main/java/seedu/address/model/filter/FilterByDatePredicate.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+public class FilterByDatePredicate implements TaskPredicate{
+	private final LocalDateTime startTime;
+	private final LocalDateTime endTime;
+
+	public FilterByDatePredicate (LocalDateTime startTime, LocalDateTime endTime){
+		assert !startTime.equals(null);
+		assert !endTime.equals(null);
+
+		this.startTime = startTime;
+		this.endTime = endTime;
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/filter/FilterByDatePredicate.java:25]: Should probably check that the startTime is not after the endTime.

File: src/main/java/seedu/address/model/filter/FilterByDatePredicate.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	@Override
+	public boolean test (DeadlineTask task){
+		if((greaterThan(task) && lesserThan(task)) | equalToMin(task) | equalToMax(task))
+			return true;
+		else
+			return false;
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/filter/FilterByDatePredicate.java:33]: Now that I have lots of time to read your PR carefully, I _think_ this is correct. But, if you wanted to get this PR merged on the deadline day itself, your code needs to be extremely easy to read so I can verify its correctness. Unfortunately, this code, with all the `greaterThan` and `lesserThan` methods, is obfuscated and hard to read.

Basically, I think what you want is this:

```
due >= start && due <= end
```

Using CS1231, we get:

```
!!(due >= start && due <= end)
!(!(due >= start) || !(due <= end)
!(due < start || due > end)
```

So this will accomplish the same thing with less code, and allows me to verify the correctness much more easily:

``` java
@Override
public boolean test(DeadlineTask task) {
    // Task due date must be within startTime and endTime inclusive
    return !(task.getDue().isBefore(startTime) || task.getDue().isAfter(endTime));
}
```

The comment tells me what the code should do, and the next line is short and succinct -- allowing me to easily verify its correctness.

This makes _much_ more sense to first time readers compared to having all of these `greaterThan`, `lesserThan` methods.

(NOTE: This is not the best approach as it does not have the best SLAP. The best SLAP would involve using our [LocalDateTimeDuration](https://github.com/CS2103AUG2016-T11-C4/main/blob/master/src/main/java/seedu/address/commons/time/LocalDateTimeDuration.java) However, this approach has minimum code changes, and would have been more likely to be merged on the deadline day.)

**Review by pyokagan — COMMENTED** [review]
@snehasp13 

Code does not pass checkstyle.

There is a fatal error in the `FilterByDatePredicateParser` that makes the command useless.

`FilterByDatePredicate` is extremely hard to read and makes it hard for me to verify its correctness under time constraints. When rushing for deadlines, I preferred to err on the side of caution.

`AddDeadlineCommandTest` does not test the command correctly.

Lots of unnecessary changes. If I had time, I would have scrubbed them for you, but I didn't at that point in time.

File: src/main/java/seedu/address/model/filter/FilterByDatePredicate.java (hunk 13 above)
Comment (Bogi1001) [reply src/main/java/seedu/address/model/filter/FilterByDatePredicate.java:33]: Thanks for your comments @pyokagan . They were quite helpful. I shall work on all the points that you have mentioned to improve my code.
